// @amplitude/experiment-node-server/dist/src/types/config.js
import { __exports as config } from "../../../../../../_virtual/config.js";
Object.defineProperty(config, "__esModule", { value: true });
config.AssignmentConfigDefaults = config.LocalEvaluationDefaults = config.Defaults = config.RemoteEvaluationDefaults = void 0;
config.RemoteEvaluationDefaults = {
  debug: false,
  serverUrl: "https://api.lab.amplitude.com",
  fetchTimeoutMillis: 1e4,
  fetchRetries: 8,
  fetchRetryBackoffMinMillis: 500,
  fetchRetryBackoffMaxMillis: 1e4,
  fetchRetryBackoffScalar: 1.5,
  fetchRetryTimeoutMillis: 1e4,
  httpAgent: null
};
config.Defaults = {
  debug: false,
  serverUrl: "https://api.lab.amplitude.com",
  fetchTimeoutMillis: 1e4,
  fetchRetries: 8,
  fetchRetryBackoffMinMillis: 500,
  fetchRetryBackoffMaxMillis: 1e4,
  fetchRetryBackoffScalar: 1.5,
  fetchRetryTimeoutMillis: 1e4,
  httpAgent: null
};
config.LocalEvaluationDefaults = {
  debug: false,
  serverUrl: "https://api.lab.amplitude.com",
  bootstrap: {},
  flagConfigPollingIntervalMillis: 3e4,
  httpAgent: null
};
config.AssignmentConfigDefaults = {
  cacheCapacity: 65536
};
export {
  config as default
};
