// @amplitude/experiment-node-server/dist/src/types/config.cjs
"use strict";
const config = require("../../../../../../_virtual/config.cjs");
Object.defineProperty(config.__exports, "__esModule", { value: true });
config.__exports.AssignmentConfigDefaults = config.__exports.LocalEvaluationDefaults = config.__exports.Defaults = config.__exports.RemoteEvaluationDefaults = void 0;
config.__exports.RemoteEvaluationDefaults = {
  debug: false,
  serverUrl: "https://api.lab.amplitude.com",
  fetchTimeoutMillis: 1e4,
  fetchRetries: 8,
  fetchRetryBackoffMinMillis: 500,
  fetchRetryBackoffMaxMillis: 1e4,
  fetchRetryBackoffScalar: 1.5,
  fetchRetryTimeoutMillis: 1e4,
  httpAgent: null
};
config.__exports.Defaults = {
  debug: false,
  serverUrl: "https://api.lab.amplitude.com",
  fetchTimeoutMillis: 1e4,
  fetchRetries: 8,
  fetchRetryBackoffMinMillis: 500,
  fetchRetryBackoffMaxMillis: 1e4,
  fetchRetryBackoffScalar: 1.5,
  fetchRetryTimeoutMillis: 1e4,
  httpAgent: null
};
config.__exports.LocalEvaluationDefaults = {
  debug: false,
  serverUrl: "https://api.lab.amplitude.com",
  bootstrap: {},
  flagConfigPollingIntervalMillis: 3e4,
  httpAgent: null
};
config.__exports.AssignmentConfigDefaults = {
  cacheCapacity: 65536
};
module.exports = config.__exports;
