// @prismicio/mocks/lib/renderers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderSliceMock = exports.renderDocumentMock = void 0;
const tslib_1 = require("tslib");
const api_renderer_1 = require("@prismicio/api-renderer");
const customtypes_1 = require("@prismicio/types-internal/lib/customtypes");
const uuid_1 = require("uuid");
const MockContext_1 = (0, tslib_1.__importDefault)(require("./MockContext"));
function renderDocumentMock(customType, sharedSlices, mocks) {
    const filteredSharedSlices = Object.entries(sharedSlices).filter((entry) => Boolean(entry[1]));
    const filteredMocks = Object.entries(mocks).filter((entry) => Boolean(entry[1]));
    const staticCustomType = (0, customtypes_1.toStatic)(customType, new Map(filteredSharedSlices));
    const customTypeDef = {
        customTypeId: customType.id,
        fields: Object.fromEntries((0, customtypes_1.flattenSections)(staticCustomType)),
    };
    return (0, api_renderer_1.DocumentRenderer)(MockContext_1.default).renderMocks(customTypeDef, filteredMocks);
}
exports.renderDocumentMock = renderDocumentMock;
function renderSliceMock(sliceModel, mock) {
    return {
        ...(0, api_renderer_1.SharedSliceRenderer)(MockContext_1.default).renderMocks(sliceModel, mock),
        id: `${sliceModel.id}$${(0, uuid_1.v4)()}`,
        slice_type: sliceModel.id,
    };
}
exports.renderSliceMock = renderSliceMock;
