// @slicemachine/manager/dist/managers/telemetry/types.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const SegmentEventType = {
  command_init_start: "command:init:start",
  command_init_identify: "command:init:identify",
  command_init_end: "command:init:end",
  review: "review",
  sliceSimulator_open: "slice-simulator:open",
  sliceSimulator_isNotRunning: "slice-simulator:is-not-running",
  pageView: "page-view",
  openVideoTutorials: "open-video-tutorials",
  field_added: "field:added",
  field_settingsOpened: "field:settings-opened",
  customType_created: "custom-type:created",
  customType_sliceZoneUpdated: "custom-type:slice-zone-updated",
  customType_openAddFromTemplates: "custom-type:open-add-from-templates",
  customType_saved: "custom-type:saved",
  slice_created: "slice:created",
  legacySlice_converted: "legacy-slice:converted",
  screenshotTaken: "screenshot-taken",
  changes_pushed: "changes:pushed",
  changes_limitReach: "changes:limit-reach",
  editor_widgetUsed: "editor:widget-used",
  open_page_snippet: "page-type:open-snippet",
  copy_page_snippet: "page-type:copy-snippet",
  switch_environment: "environment:switch",
  devCollab_joinBetaClicked: "dev-collab:join-beta-clicked",
  devCollab_setUpWorkflowOpened: "dev-collab:set-up-workflow-opened",
  devCollab_workflowStubDisplayed: "dev-collab:workflow-stub-displayed",
  sliceMachine_start: "slice-machine:start",
  sliceLibrary_beta_modalOpened: "slice-library:beta:modal-opened",
  sliceLibrary_beta_codeOpened: "slice-library:beta:code-opened"
};
const HumanSegmentEventType = {
  [SegmentEventType.command_init_start]: "SliceMachine Init Start",
  [SegmentEventType.command_init_identify]: "SliceMachine Init Identify",
  [SegmentEventType.command_init_end]: "SliceMachine Init End",
  [SegmentEventType.review]: "SliceMachine Review",
  [SegmentEventType.sliceSimulator_open]: "SliceMachine Slice Simulator Open",
  [SegmentEventType.sliceSimulator_isNotRunning]: "SliceMachine Slice Simulator is not running",
  [SegmentEventType.pageView]: "SliceMachine Page View",
  [SegmentEventType.openVideoTutorials]: "SliceMachine Open Video Tutorials",
  [SegmentEventType.field_added]: "SliceMachine Field Added",
  [SegmentEventType.field_settingsOpened]: "SliceMachine Field Settings Opened",
  [SegmentEventType.customType_created]: "SliceMachine Custom Type Created",
  [SegmentEventType.customType_sliceZoneUpdated]: "SliceMachine Slicezone Updated",
  [SegmentEventType.customType_openAddFromTemplates]: "SliceMachine Open Add from templates",
  [SegmentEventType.customType_saved]: "SliceMachine Custom Type Saved",
  [SegmentEventType.slice_created]: "SliceMachine Slice Created",
  [SegmentEventType.legacySlice_converted]: "SliceMachine Legacy Slice Converted",
  [SegmentEventType.screenshotTaken]: "SliceMachine Screenshot Taken",
  [SegmentEventType.changes_pushed]: "SliceMachine Changes Pushed",
  [SegmentEventType.changes_limitReach]: "SliceMachine Changes Limit Reach",
  [SegmentEventType.editor_widgetUsed]: "SliceMachine Editor Widget Used",
  [SegmentEventType.open_page_snippet]: "SliceMachine Opens Page Type Snippet Dialog",
  [SegmentEventType.copy_page_snippet]: "Slice Machine page code snippet copied",
  [SegmentEventType.switch_environment]: "SliceMachine environment switch",
  [SegmentEventType.devCollab_joinBetaClicked]: "SliceMachine Dev Collab Join Beta Clicked",
  [SegmentEventType.devCollab_setUpWorkflowOpened]: "SliceMachine Dev Collab Set Up Workflow Opened",
  [SegmentEventType.devCollab_workflowStubDisplayed]: "SliceMachine Dev Collab Workflow Stub Displayed",
  [SegmentEventType.sliceMachine_start]: "SliceMachine Start",
  [SegmentEventType.sliceLibrary_beta_modalOpened]: "SliceMachine Slice Library [BETA] CTA modal displayed",
  [SegmentEventType.sliceLibrary_beta_codeOpened]: "SliceMachine Slice Library [BETA] CTA example code opened"
};
exports.HumanSegmentEventType = HumanSegmentEventType;
exports.SegmentEventType = SegmentEventType;
