// @amplitude/experiment-node-server/dist/src/local/fetcher.js
import { commonjsGlobal } from "../../../../../../_virtual/_commonjsHelpers.js";
import { __exports as fetcher } from "../../../../../../_virtual/fetcher.js";
import require$$1 from "../../../../../../_virtual/experiment-core.esm.js";
import "../../gen/version.js";
import "../transport/http.js";
import "../types/config.js";
import "../util/logger.js";
import { __exports as version } from "../../../../../../_virtual/version.js";
import { __exports as http } from "../../../../../../_virtual/http.js";
import { __exports as config } from "../../../../../../_virtual/config.js";
import { __exports as logger } from "../../../../../../_virtual/logger.js";
var __awaiter = commonjsGlobal && commonjsGlobal.__awaiter || function(thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function(resolve) {
      resolve(value);
    });
  }
  return new (P || (P = Promise))(function(resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }
    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }
    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }
    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};
Object.defineProperty(fetcher, "__esModule", { value: true });
fetcher.FlagConfigFetcher = void 0;
const experiment_core_1 = require$$1;
const version_1 = version;
const http_1 = http;
const config_1 = config;
const logger_1 = logger;
const FLAG_CONFIG_TIMEOUT = 5e3;
class FlagConfigFetcher {
  constructor(apiKey, httpClient, serverUrl = config_1.LocalEvaluationDefaults.serverUrl, debug = false) {
    this.apiKey = apiKey;
    this.serverUrl = serverUrl;
    this.flagApi = new experiment_core_1.SdkFlagApi(apiKey, serverUrl, new http_1.WrapperClient(httpClient));
    this.logger = new logger_1.ConsoleLogger(debug);
  }
  /**
   * Fetch local evaluation mode flag configs from the Experiment API server.
   * These flag configs can be used to perform local evaluation.
   *
   * @returns The local evaluation mode flag configs for the initialized
   * environment
   */
  fetch() {
    return __awaiter(this, void 0, void 0, function* () {
      const flags = this.flagApi.getFlags({
        libraryName: "experiment-node-server",
        libraryVersion: version_1.version,
        evaluationMode: "local",
        timeoutMillis: FLAG_CONFIG_TIMEOUT
      });
      if (this.receiver) {
        this.receiver(JSON.stringify(flags));
      }
      return flags;
    });
  }
  setRawReceiver(rawReceiver) {
    this.receiver = rawReceiver;
  }
}
fetcher.FlagConfigFetcher = FlagConfigFetcher;
export {
  fetcher as default
};
