// @prismicio/types-internal/lib/content/fields/nestable/FieldContent/RangeContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RangeContent = exports.RangeLegacy = exports.isRangeContent = void 0;
const tslib_1 = require("tslib");
const fp_ts_1 = require("fp-ts");
const function_1 = require("fp-ts/lib/function");
const t = (0, tslib_1.__importStar)(require("io-ts"));
const widgets_1 = require("../../../../customtypes/widgets");
const TrimmedString_1 = (0, tslib_1.__importDefault)(require("../../../../validators/TrimmedString"));
const utils_1 = require("../../../utils");
const common_1 = require("./common");
const isRangeContent = (u) => (0, utils_1.hasContentType)(u) &&
    u.__TYPE__ === common_1.FieldContentType &&
    (0, utils_1.hasFieldContentType)(u) &&
    u.type === "Range";
exports.isRangeContent = isRangeContent;
const RangeLegacy = (ctx) => new t.Type("RangeLegacy", exports.isRangeContent, (u) => {
    return (0, function_1.pipe)(TrimmedString_1.default.decode(u), fp_ts_1.either.map((f) => exports.RangeContent.encode({
        value: f,
        type: widgets_1.RangeFieldType,
        __TYPE__: common_1.FieldContentType,
    })));
}, (b) => {
    return {
        content: b.value,
        types: { [ctx.keyOfType]: widgets_1.RangeFieldType },
    };
});
exports.RangeLegacy = RangeLegacy;
exports.RangeContent = t.strict({
    type: t.literal(widgets_1.RangeFieldType),
    value: TrimmedString_1.default,
    __TYPE__: t.literal(common_1.FieldContentType),
});
