// @amplitude/analytics-node/lib/esm/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const nodeClient = require("./node-client.cjs");
const index = require("../../../analytics-types/lib/esm/index.cjs");
const revenue$1 = require("../../../analytics-core/lib/esm/revenue.cjs");
const identify$1 = require("../../../analytics-core/lib/esm/identify.cjs");
var add = nodeClient.default.add, groupIdentify = nodeClient.default.groupIdentify, identify = nodeClient.default.identify, init = nodeClient.default.init, logEvent = nodeClient.default.logEvent, remove = nodeClient.default.remove, revenue = nodeClient.default.revenue, setGroup = nodeClient.default.setGroup, setOptOut = nodeClient.default.setOptOut, track = nodeClient.default.track, flush = nodeClient.default.flush;
exports.createInstance = nodeClient.createInstance;
exports.Types = index;
exports.Revenue = revenue$1.Revenue;
exports.Identify = identify$1.Identify;
exports.add = add;
exports.flush = flush;
exports.groupIdentify = groupIdentify;
exports.identify = identify;
exports.init = init;
exports.logEvent = logEvent;
exports.remove = remove;
exports.revenue = revenue;
exports.setGroup = setGroup;
exports.setOptOut = setOptOut;
exports.track = track;
