// @amplitude/analytics-core/lib/esm/revenue.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const tslib_es6 = require("../../../../tslib/tslib.es6.cjs");
const validProperties = require("./utils/valid-properties.cjs");
const event = require("../../../analytics-types/lib/esm/event.cjs");
var Revenue = (
  /** @class */
  function() {
    function Revenue2() {
      this.productId = "";
      this.quantity = 1;
      this.price = 0;
    }
    Revenue2.prototype.setProductId = function(productId) {
      this.productId = productId;
      return this;
    };
    Revenue2.prototype.setQuantity = function(quantity) {
      if (quantity > 0) {
        this.quantity = quantity;
      }
      return this;
    };
    Revenue2.prototype.setPrice = function(price) {
      this.price = price;
      return this;
    };
    Revenue2.prototype.setRevenueType = function(revenueType) {
      this.revenueType = revenueType;
      return this;
    };
    Revenue2.prototype.setRevenue = function(revenue) {
      this.revenue = revenue;
      return this;
    };
    Revenue2.prototype.setEventProperties = function(properties) {
      if (validProperties.isValidObject(properties)) {
        this.properties = properties;
      }
      return this;
    };
    Revenue2.prototype.getEventProperties = function() {
      var eventProperties = this.properties ? tslib_es6.__assign({}, this.properties) : {};
      eventProperties[event.RevenueProperty.REVENUE_PRODUCT_ID] = this.productId;
      eventProperties[event.RevenueProperty.REVENUE_QUANTITY] = this.quantity;
      eventProperties[event.RevenueProperty.REVENUE_PRICE] = this.price;
      eventProperties[event.RevenueProperty.REVENUE_TYPE] = this.revenueType;
      eventProperties[event.RevenueProperty.REVENUE] = this.revenue;
      return eventProperties;
    };
    return Revenue2;
  }()
);
exports.Revenue = Revenue;
