// @slicemachine/manager/dist/managers/createSliceMachineManagerMiddleware.js
import { createRPCMiddleware } from './../_node_modules/r19/dist/createRPCMiddleware.js';
import { proceduresFromInstance } from './../_node_modules/r19/dist/proceduresFromInstance.js';
const defineOmits = () => {
  return (omit) => {
    return omit;
  };
};
const omitProcedures = defineOmits()([
  "_sliceMachinePluginRunner",
  "_prismicAuthManager",
  "customTypes._sliceMachineManager",
  "plugins._sliceMachineManager",
  "prismicRepository._sliceMachineManager",
  "project._sliceMachineManager",
  "screenshots._sliceMachineManager",
  "simulator._sliceMachineManager",
  "slices._sliceMachineManager",
  "snippets._sliceMachineManager",
  "telemetry._sliceMachineManager",
  "user._sliceMachineManager",
  "versions._sliceMachineManager",
  "documentation._sliceMachineManager",
  "sliceTemplateLibrary._sliceMachineManager",
  "git._sliceMachineManager",
  "getSliceMachinePluginRunner",
  "getPrismicAuthManager"
]);
const getSliceMachineManagerProcedures = (args) => {
  return proceduresFromInstance(args.sliceMachineManager, {
    omit: omitProcedures
  });
};
const createSliceMachineManagerMiddleware = (args) => {
  return createRPCMiddleware({
    procedures: getSliceMachineManagerProcedures({
      sliceMachineManager: args.sliceMachineManager
    }),
    onError: args.onError
  });
};
export {
  createSliceMachineManagerMiddleware,
  getSliceMachineManagerProcedures
};
