// @slicemachine/plugin-kit/dist/createSliceMachinePluginRunner.js
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
import * as path from "node:path";
import { createRequire } from "node:module";
import { defu } from "defu";
import { createSliceMachineContext } from "./createSliceMachineContext.js";
import { createSliceMachineHookSystem } from "./createSliceMachineHookSystem.js";
import { createSliceMachineActions } from "./createSliceMachineActions.js";
import { createSliceMachineHelpers } from "./createSliceMachineHelpers.js";
const REQUIRED_ADAPTER_HOOKS = [
  "slice:create",
  "slice:read",
  "slice:rename",
  "slice:delete",
  "slice:update",
  "slice:asset:update",
  "slice:asset:read",
  "slice:asset:delete",
  "slice-library:read",
  "custom-type:create",
  "custom-type:read",
  "custom-type:rename",
  "custom-type:delete",
  "custom-type:update",
  "custom-type:asset:update",
  "custom-type:asset:read",
  "custom-type:asset:delete",
  "custom-type-library:read"
];
const ADAPTER_ONLY_HOOKS = [
  "slice:read",
  "slice:asset:read",
  "slice-library:read",
  "custom-type:read",
  "custom-type:asset:read",
  "custom-type-library:read",
  "project:environment:read"
];
class SliceMachinePluginRunner {
  constructor({ project, hookSystem, nativePlugins = {} }) {
    __publicField(this, "_project");
    __publicField(this, "_hookSystem");
    __publicField(this, "_nativePlugins");
    /**
     * Slice Machine actions provided to hooks.
     *
     * IMPORTANT: Prefer creating your own abstraction over using `rawActions`
     * directly to prevent code breakage if this internal API changes.
     *
     * @internal
     */
    __publicField(this, "rawActions");
    /**
     * Slice Machine helpers provided to hooks.
     *
     * IMPORTANT: Prefer creating your own abstraction over using `rawHelpers`
     * directly to prevent code breakage if this internal API changes.
     *
     * @internal
     */
    __publicField(this, "rawHelpers");
    // Methods forwarded to the plugin runner's hook system.
    __publicField(this, "callHook");
    __publicField(this, "hooksForOwner");
    __publicField(this, "hooksForType");
    __publicField(this, "createScope");
    this._project = project;
    this._hookSystem = hookSystem;
    this._nativePlugins = nativePlugins;
    this.rawActions = createSliceMachineActions(this._project, this._hookSystem);
    this.rawHelpers = createSliceMachineHelpers(this._project);
    this.callHook = this._hookSystem.callHook.bind(this._hookSystem);
    this.hooksForOwner = this._hookSystem.hooksForOwner.bind(this._hookSystem);
    this.hooksForType = this._hookSystem.hooksForType.bind(this._hookSystem);
    this.createScope = this._hookSystem.createScope.bind(this._hookSystem);
  }
  async _loadPlugin(pluginRegistration) {
    const { resolve, options = {} } = typeof pluginRegistration === "object" && "resolve" in pluginRegistration ? pluginRegistration : { resolve: pluginRegistration };
    let plugin = void 0;
    if (typeof resolve === "string") {
      try {
        const raw = await createRequire(path.resolve(this._project.root, "noop.js"))(resolve);
        plugin = raw.default || raw;
      } catch (error) {
      }
      if (!plugin) {
        plugin = this._nativePlugins[resolve];
      }
      if (!plugin) {
        throw new Error(`Could not resolve plugin \`${resolve}\`. Check that it has been installed.`);
      }
    } else {
      plugin = resolve;
    }
    if (!plugin) {
      throw new Error(`Could not load plugin: \`${resolve}\``);
    }
    const mergedOptions = defu(options, plugin.defaultOptions || {});
    return {
      ...plugin,
      resolve,
      options: mergedOptions
    };
  }
  async _setupPlugin(plugin, as) {
    const context = createSliceMachineContext({
      actions: this.rawActions,
      helpers: this.rawHelpers,
      project: this._project,
      plugin
    });
    const hookSystemScope = this._hookSystem.createScope(plugin.meta.name, [context]);
    const hook = as === "adapter" ? hookSystemScope.hook : (type, hook2, ...args) => {
      if (ADAPTER_ONLY_HOOKS.includes(type)) {
        return;
      }
      return hookSystemScope.hook(type, hook2, ...args);
    };
    try {
      await plugin.setup({
        ...context,
        hook,
        unhook: hookSystemScope.unhook
      });
    } catch (error) {
      if (error instanceof Error) {
        throw new Error(`Plugin \`${plugin.meta.name}\` errored during setup: ${error.message}`, { cause: error });
      } else {
        throw new Error(`Plugin \`${plugin.meta.name}\` errored during setup: ${error}`);
      }
    }
  }
  _validateAdapter(adapter) {
    const hooks = this._hookSystem.hooksForOwner(adapter.meta.name);
    const hookTypes = hooks.map((hook) => hook.meta.type);
    const missingHooks = REQUIRED_ADAPTER_HOOKS.filter((requiredHookType) => !hookTypes.includes(requiredHookType));
    if (missingHooks.length) {
      throw new Error(`Adapter \`${adapter.meta.name}\` is missing hooks: \`${missingHooks.join("`, `")}\``);
    }
  }
  async init() {
    const [adapter, ...plugins] = await Promise.all([
      this._project.config.adapter,
      ...this._project.config.plugins ?? []
    ].map((pluginRegistration) => this._loadPlugin(pluginRegistration)));
    await Promise.all([
      this._setupPlugin(adapter, "adapter"),
      ...plugins.map((plugin) => this._setupPlugin(plugin, "plugin"))
    ]);
    this._validateAdapter(adapter);
  }
}
const createSliceMachinePluginRunner = ({ project, nativePlugins }) => {
  const hookSystem = createSliceMachineHookSystem();
  return new SliceMachinePluginRunner({
    project,
    hookSystem,
    nativePlugins
  });
};
export {
  ADAPTER_ONLY_HOOKS,
  REQUIRED_ADAPTER_HOOKS,
  SliceMachinePluginRunner,
  createSliceMachinePluginRunner
};
