// @slicemachine/plugin-kit/dist/types.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const SliceMachineHookType = {
  slice_create: "slice:create",
  slice_update: "slice:update",
  slice_rename: "slice:rename",
  slice_delete: "slice:delete",
  slice_read: "slice:read",
  slice_asset_update: "slice:asset:update",
  slice_asset_delete: "slice:asset:delete",
  slice_asset_read: "slice:asset:read",
  sliceLibrary_read: "slice-library:read",
  customType_create: "custom-type:create",
  customType_update: "custom-type:update",
  customType_rename: "custom-type:rename",
  customType_delete: "custom-type:delete",
  customType_read: "custom-type:read",
  customType_asset_update: "custom-type:asset:update",
  customType_asset_delete: "custom-type:asset:delete",
  customType_asset_read: "custom-type:asset:read",
  customTypeLibrary_read: "custom-type-library:read",
  documentation_read: "documentation:read",
  sliceTemplateLibrary_read: "slice-template-library:read",
  snippet_read: "snippet:read",
  project_init: "project:init",
  project_environment_read: "project:environment:read",
  project_environment_update: "project:environment:update",
  debug: "debug"
};
exports.SliceMachineHookType = SliceMachineHookType;
