// @prismicio/types-internal/src/content/fields/nestable/FieldContent/ColorContent.ts
import { either } from "fp-ts"
import { pipe } from "fp-ts/lib/function"
import * as t from "io-ts"

import { ColorFieldType } from "../../../../customtypes/widgets"
import TrimmedString from "../../../../validators/TrimmedString"
import type { LegacyContentCtx, WithTypes } from "../../../LegacyContentCtx"
import { hasContentType, hasFieldContentType } from "../../../utils"
import { FieldContentType } from "./common"

export const isColorContent = (u: unknown): u is ColorContent =>
	hasContentType(u) &&
	u.__TYPE__ === FieldContentType &&
	hasFieldContentType(u) &&
	u.type === "Color"

export const ColorLegacy = (ctx: LegacyContentCtx) =>
	new t.Type<ColorContent, WithTypes<string>, unknown>(
		"ColorLegacy",
		isColorContent,
		(u) => {
			return pipe(
				TrimmedString.decode(u),
				either.map((f) =>
					ColorContent.encode({
						value: f,
						type: ColorFieldType,
						__TYPE__: FieldContentType,
					}),
				),
			)
		},

		(b: ColorContent) => {
			return {
				content: b.value,
				types: { [ctx.keyOfType]: ColorFieldType },
			}
		},
	)

export const ColorContent = t.strict({
	type: t.literal(ColorFieldType),
	value: TrimmedString,
	__TYPE__: t.literal(FieldContentType),
})

export type ColorContent = t.TypeOf<typeof ColorContent>
