// @prismicio/types-internal/src/content/fields/nestable/FieldContent/DateContent.ts
import { either } from "fp-ts"
import { pipe } from "fp-ts/lib/function"
import * as t from "io-ts"

import { DateFieldType } from "../../../../customtypes/widgets"
import TrimmedString from "../../../../validators/TrimmedString"
import type { LegacyContentCtx, WithTypes } from "../../../LegacyContentCtx"
import { hasContentType, hasFieldContentType } from "../../../utils"
import { FieldContentType } from "./common"

export const isDateContent = (u: unknown): u is DateContent =>
	hasContentType(u) &&
	u.__TYPE__ === FieldContentType &&
	hasFieldContentType(u) &&
	u.type === "Date"

export const DateLegacy = (ctx: LegacyContentCtx) =>
	new t.Type<DateContent, WithTypes<string>, unknown>(
		"DateLegacy",
		isDateContent,
		(u) => {
			return pipe(
				TrimmedString.decode(u),
				either.map((f) =>
					DateContent.encode({
						value: f,
						type: DateFieldType,
						__TYPE__: FieldContentType,
					}),
				),
			)
		},

		(b: DateContent) => {
			return {
				content: b.value,
				types: { [ctx.keyOfType]: DateFieldType },
			}
		},
	)

export const DateContent = t.strict({
	type: t.literal(DateFieldType),
	value: TrimmedString,
	__TYPE__: t.literal(FieldContentType),
})

export type DateContent = t.TypeOf<typeof DateContent>
