// @prismicio/types-internal/src/content/fields/nestable/FieldContent/index.ts
import * as t from "io-ts"

import { ColorContent } from "./ColorContent"
import { DateContent } from "./DateContent"
import { NumberContent } from "./NumberContent"
import { RangeContent } from "./RangeContent"
import { SelectContent } from "./SelectContent"
import { TextContent } from "./TextContent"
import { TimestampContent } from "./TimestampContent"

export const FieldType = t.keyof({
	Text: null,
	Date: null,
	Timestamp: null,
	Color: null,
	Number: null,
	Range: null,
	Select: null,
})
export type FieldType = t.TypeOf<typeof FieldType>

export const FieldContent = t.union([
	DateContent,
	ColorContent,
	NumberContent,
	RangeContent,
	SelectContent,
	TextContent,
	TimestampContent,
])
export type FieldContent = t.TypeOf<typeof FieldContent>

export * from "./ColorContent"
export * from "./common"
export * from "./DateContent"
export * from "./NumberContent"
export * from "./RangeContent"
export * from "./SelectContent"
export * from "./TextContent"
export * from "./TimestampContent"
