// @prismicio/types-internal/src/content/fields/nestable/FieldContent/NumberContent.ts
import { either } from "fp-ts"
import { pipe } from "fp-ts/lib/function"
import * as t from "io-ts"

import { NumberFieldType } from "../../../../customtypes/widgets"
import TrimmedString from "../../../../validators/TrimmedString"
import type { LegacyContentCtx, WithTypes } from "../../../LegacyContentCtx"
import { hasContentType, hasFieldContentType } from "../../../utils"
import { FieldContentType } from "./common"

export const isNumberContent = (u: unknown): u is NumberContent =>
	hasContentType(u) &&
	u.__TYPE__ === FieldContentType &&
	hasFieldContentType(u) &&
	u.type === "Number"

export const NumberLegacy = (ctx: LegacyContentCtx) =>
	new t.Type<NumberContent, WithTypes<string>, unknown>(
		"NumberLegacy",
		isNumberContent,
		(u) => {
			return pipe(
				TrimmedString.decode(u),
				either.map((f) =>
					NumberContent.encode({
						value: f,
						type: NumberFieldType,
						__TYPE__: FieldContentType,
					}),
				),
			)
		},

		(b: NumberContent) => {
			return {
				content: b.value,
				types: { [ctx.keyOfType]: NumberFieldType },
			}
		},
	)

export const NumberContent = t.strict({
	type: t.literal(NumberFieldType),
	value: TrimmedString,
	__TYPE__: t.literal(FieldContentType),
})

export type NumberContent = t.TypeOf<typeof NumberContent>
