// @prismicio/types-internal/src/content/fields/nestable/FieldContent/RangeContent.ts
import { either } from "fp-ts"
import { pipe } from "fp-ts/lib/function"
import * as t from "io-ts"

import { RangeFieldType } from "../../../../customtypes/widgets"
import TrimmedString from "../../../../validators/TrimmedString"
import type { LegacyContentCtx, WithTypes } from "../../../LegacyContentCtx"
import { hasContentType, hasFieldContentType } from "../../../utils"
import { FieldContentType } from "./common"

export const isRangeContent = (u: unknown): u is RangeContent =>
	hasContentType(u) &&
	u.__TYPE__ === FieldContentType &&
	hasFieldContentType(u) &&
	u.type === "Range"

export const RangeLegacy = (ctx: LegacyContentCtx) =>
	new t.Type<RangeContent, WithTypes<string>, unknown>(
		"RangeLegacy",
		isRangeContent,
		(u) => {
			return pipe(
				TrimmedString.decode(u),
				either.map((f) =>
					RangeContent.encode({
						value: f,
						type: RangeFieldType,
						__TYPE__: FieldContentType,
					}),
				),
			)
		},

		(b: RangeContent) => {
			return {
				content: b.value,
				types: { [ctx.keyOfType]: RangeFieldType },
			}
		},
	)

export const RangeContent = t.strict({
	type: t.literal(RangeFieldType),
	value: TrimmedString,
	__TYPE__: t.literal(FieldContentType),
})

export type RangeContent = t.TypeOf<typeof RangeContent>
