// @prismicio/types-internal/src/content/fields/nestable/FieldContent/TextContent.ts
import { either } from "fp-ts"
import { pipe } from "fp-ts/lib/function"
import * as t from "io-ts"

import { TextFieldType } from "../../../../customtypes/widgets"
import TrimmedString from "../../../../validators/TrimmedString"
import type { LegacyContentCtx, WithTypes } from "../../../LegacyContentCtx"
import { hasContentType, hasFieldContentType } from "../../../utils"
import { FieldContentType } from "./common"

export const isTextContent = (u: unknown): u is TextContent =>
	hasContentType(u) &&
	u.__TYPE__ === FieldContentType &&
	hasFieldContentType(u) &&
	u.type === "Text"

export const TextLegacy = (ctx: LegacyContentCtx) =>
	new t.Type<TextContent, WithTypes<string>, unknown>(
		"TextLegacy",
		isTextContent,
		(u) => {
			return pipe(
				TrimmedString.decode(u),
				either.map((f) =>
					TextContent.encode({
						value: f,
						type: TextFieldType,
						__TYPE__: FieldContentType,
					}),
				),
			)
		},

		(b: TextContent) => {
			return {
				content: b.value,
				types: { [ctx.keyOfType]: TextFieldType },
			}
		},
	)

export const TextContent = t.strict({
	type: t.literal(TextFieldType),
	value: TrimmedString,
	__TYPE__: t.literal(FieldContentType),
})

export type TextContent = t.TypeOf<typeof TextContent>
