// @prismicio/types-internal/src/content/fields/nestable/FieldContent/TimestampContent.ts
import { either } from "fp-ts"
import { pipe } from "fp-ts/lib/function"
import * as t from "io-ts"

import { TimestampFieldType } from "../../../../customtypes/widgets"
import TrimmedString from "../../../../validators/TrimmedString"
import type { LegacyContentCtx, WithTypes } from "../../../LegacyContentCtx"
import { hasContentType, hasFieldContentType } from "../../../utils"
import { FieldContentType } from "./common"

export const isTimestampContent = (u: unknown): u is TimestampContent =>
	hasContentType(u) &&
	u.__TYPE__ === FieldContentType &&
	hasFieldContentType(u) &&
	u.type === "Timestamp"

export const TimestampLegacy = (ctx: LegacyContentCtx) =>
	new t.Type<TimestampContent, WithTypes<string>, unknown>(
		"TimestampLegacy",
		isTimestampContent,
		(u) => {
			return pipe(
				TrimmedString.decode(u),
				either.map((f) =>
					TimestampContent.encode({
						value: f,
						type: TimestampFieldType,
						__TYPE__: FieldContentType,
					}),
				),
			)
		},

		(b: TimestampContent) => {
			return {
				content: b.value,
				types: { [ctx.keyOfType]: TimestampFieldType },
			}
		},
	)

export const TimestampContent = t.strict({
	type: t.literal(TimestampFieldType),
	value: TrimmedString,
	__TYPE__: t.literal(FieldContentType),
})

export type TimestampContent = t.TypeOf<typeof TimestampContent>
