// start-slicemachine/dist/legacyMigrations/detectFramework.js
import { readFile } from "node:fs/promises";
import { join } from "node:path";
import semver from './../_node_modules/semver/index.js';
const FRAMEWORKS = {
  previousNuxt: {
    name: "Nuxt 2 (legacy)",
    adapterName: "@slicemachine/adapter-nuxt2",
    adapterVersion: "latest",
    runProjectInitHook: true,
    compatibility: {
      nuxt: "^2.0.0",
      "nuxt-sm": "*"
    }
  },
  nuxt: {
    name: "Nuxt 2",
    adapterName: "@slicemachine/adapter-nuxt2",
    adapterVersion: "latest",
    runProjectInitHook: false,
    compatibility: {
      nuxt: "^2.0.0"
    }
  },
  previousNext: {
    name: "Next.js (legacy)",
    adapterName: "@slicemachine/adapter-next",
    adapterVersion: "latest",
    runProjectInitHook: true,
    compatibility: {
      next: ">=11.0.0",
      "prismic-reactjs": "*"
    }
  },
  next: {
    name: "Next.js",
    adapterName: "@slicemachine/adapter-next",
    adapterVersion: "latest",
    runProjectInitHook: false,
    compatibility: {
      next: ">=11.0.0"
    }
  }
};
const UNIVERSAL = {
  name: "universal (no framework)",
  adapterName: "@slicemachine/adapter-universal",
  adapterVersion: "latest",
  runProjectInitHook: false,
  compatibility: {}
};
const detectFramework = async (cwd) => {
  const path = join(cwd, "package.json");
  let allDependencies;
  try {
    const pkg = JSON.parse(await readFile(path, "utf-8"));
    allDependencies = {
      ...pkg.dependencies,
      ...pkg.devDependencies
    };
  } catch (error) {
    throw new Error(`Failed to read project's \`package.json\` at \`${path}\``, { cause: error });
  }
  return Object.values(FRAMEWORKS).find((framework) => {
    return Object.entries(framework.compatibility).every(([pkg, range]) => {
      if (pkg in allDependencies) {
        try {
          const minimumVersion = semver.minVersion(allDependencies[pkg]);
          return semver.satisfies(minimumVersion, range);
        } catch (error) {
          return true;
        }
      }
      return false;
    });
  }) || UNIVERSAL;
};
export {
  FRAMEWORKS,
  UNIVERSAL,
  detectFramework
};
