// start-slicemachine/dist/legacyMigrations/migrateSMJSON.js
import * as path from "node:path";
import * as fs from "node:fs/promises";
import chalk from "chalk";
import semver from './../_node_modules/semver/index.js';
import { locateFileUpward } from "../lib/locateFileUpward.js";
import { FRAMEWORKS, UNIVERSAL, detectFramework } from "./detectFramework.js";
const migrateSMJSON = async (manager) => {
  let smJSONPath;
  try {
    smJSONPath = await locateFileUpward("sm.json", { startDir: manager.cwd });
  } catch (error) {
  }
  if (!smJSONPath) {
    return;
  }
  const smJSON = JSON.parse(await fs.readFile(smJSONPath, "utf-8"));
  if ("//" in smJSON && !("libraries" in smJSON)) {
    return console.log(`
${chalk.bgYellow(` ${chalk.black("WARN")} `)} ${chalk.magenta("sm.json")} is deprecated!

  - If you were importing values from this file, please now import
    them from ${chalk.magenta("slicemachine.config.json")}.
    You can safely delete this file after to suppress this warning.

  - If you weren't, you can safely delete this file to suppress
    this warning.
`);
  }
  if (smJSON._latest && !semver.satisfies(smJSON._latest, ">0.3.0")) {
    console.log(`
${chalk.bgYellow(` ${chalk.black("WARN")} `)} ${chalk.magenta("sm.json")} was last migrated before ${chalk.magenta("0.3.0")}, migration might be incomplete`);
  }
  console.log(`
${chalk.bgCyan(` ${chalk.black("INFO")} `)} Detected deprecated ${chalk.magenta("sm.json")}, beginning migration...`);
  let framework;
  if (smJSON.framework) {
    framework = FRAMEWORKS[smJSON.framework] || UNIVERSAL;
  } else {
    framework = await detectFramework(path.dirname(smJSONPath));
  }
  const libraries = smJSON.libraries.map((library) => {
    if (/^[@~]\//.test(library)) {
      return library.replace(/^[@~]\//, "./");
    }
    return library;
  }) || [];
  const sliceMachineConfig = {
    apiEndpoint: smJSON.apiEndpoint,
    // Infer repository name from API endpoint
    repositoryName: new URL(smJSON.apiEndpoint).host.split(".")[0],
    adapter: framework.adapterName,
    libraries
  };
  if (smJSON.localSliceSimulatorURL) {
    sliceMachineConfig.localSliceSimulatorURL = smJSON.localSliceSimulatorURL;
  }
  const sliceMachineConfigPath = await manager.project.suggestSliceMachineConfigPath();
  await manager.project.writeSliceMachineConfig({
    config: sliceMachineConfig,
    path: sliceMachineConfigPath
  });
  await fs.writeFile(smJSONPath, JSON.stringify({
    "//": "sm.json is deprecated, if you were importing values from this file please now import them from slicemachine.config.json, else you can safely delete this file",
    // We keep API endpoint in it because users might have been importing it from this file.
    apiEndpoint: smJSON.apiEndpoint
  }, void 0, "	"));
  try {
    const { execaProcess } = await manager.project.installDependencies({
      dependencies: {
        [framework.adapterName]: framework.adapterVersion
      },
      dev: true,
      log: () => {
      }
    });
    await execaProcess;
  } catch (error) {
    if (error instanceof Error && "shortMessage" in error && "stderr" in error) {
      throw new Error(`${error.shortMessage}

${error.stderr}`, {
        cause: error
      });
    }
    throw error;
  }
  if (framework.runProjectInitHook) {
    await manager.plugins.initPlugins();
    await manager.project.initProject({
      log: () => {
      }
    });
  }
  return console.log(`
${chalk.bgCyan(` ${chalk.black("INFO")} `)} Your ${chalk.magenta("sm.json")} content was automatically migrated to ${chalk.magenta("slicemachine.config.json")}

  - If you were importing values from this file, please now import
    them from ${chalk.magenta("slicemachine.config.json")}.
    You can safely delete this file after.

  - If you weren't, you can safely delete this file.
`);
};
export {
  migrateSMJSON
};
