// @sentry/node/esm/integrations/modules.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const fs = require("fs");
const path = require("path");
const integration = require('./../../_node_modules/@sentry/core/esm/integration.cjs');
let moduleCache;
const INTEGRATION_NAME = "Modules";
function getPaths() {
  try {
    return require.cache ? Object.keys(require.cache) : [];
  } catch (e) {
    return [];
  }
}
function collectModules() {
  const mainPaths = require.main && require.main.paths || [];
  const paths = getPaths();
  const infos = {};
  const seen = {};
  paths.forEach((path$1) => {
    let dir = path$1;
    const updir = () => {
      const orig = dir;
      dir = path.dirname(orig);
      if (!dir || orig === dir || seen[orig]) {
        return void 0;
      }
      if (mainPaths.indexOf(dir) < 0) {
        return updir();
      }
      const pkgfile = path.join(orig, "package.json");
      seen[orig] = true;
      if (!fs.existsSync(pkgfile)) {
        return updir();
      }
      try {
        const info = JSON.parse(fs.readFileSync(pkgfile, "utf8"));
        infos[info.name] = info.version;
      } catch (_oO) {
      }
    };
    updir();
  });
  return infos;
}
function _getModules() {
  if (!moduleCache) {
    moduleCache = collectModules();
  }
  return moduleCache;
}
const _modulesIntegration = () => {
  return {
    name: INTEGRATION_NAME,
    // TODO v8: Remove this
    setupOnce() {
    },
    // eslint-disable-line @typescript-eslint/no-empty-function
    processEvent(event) {
      event.modules = {
        ...event.modules,
        ..._getModules()
      };
      return event;
    }
  };
};
const modulesIntegration = integration.defineIntegration(_modulesIntegration);
integration.convertIntegrationFnToClass(INTEGRATION_NAME, modulesIntegration);
exports.modulesIntegration = modulesIntegration;
