// @sentry/node/esm/integrations/spotlight.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const http = require("http");
const url = require("url");
const integration = require('./../../_node_modules/@sentry/core/esm/integration.cjs');
const logger = require('./../../_node_modules/@sentry/utils/esm/logger.cjs');
const envelope = require('./../../_node_modules/@sentry/utils/esm/envelope.cjs');
function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const http__namespace = /* @__PURE__ */ _interopNamespaceDefault(http);
const INTEGRATION_NAME = "Spotlight";
const _spotlightIntegration = (options = {}) => {
  const _options = {
    sidecarUrl: options.sidecarUrl || "http://localhost:8969/stream"
  };
  return {
    name: INTEGRATION_NAME,
    // TODO v8: Remove this
    setupOnce() {
    },
    // eslint-disable-line @typescript-eslint/no-empty-function
    setup(client) {
      if (typeof process === "object" && process.env && process.env.NODE_ENV !== "development") {
        logger.logger.warn("[Spotlight] It seems you're not in dev mode. Do you really want to have Spotlight enabled?");
      }
      connectToSpotlight(client, _options);
    }
  };
};
const spotlightIntegration = integration.defineIntegration(_spotlightIntegration);
integration.convertIntegrationFnToClass(INTEGRATION_NAME, spotlightIntegration);
function connectToSpotlight(client, options) {
  const spotlightUrl = parseSidecarUrl(options.sidecarUrl);
  if (!spotlightUrl) {
    return;
  }
  let failedRequests = 0;
  if (typeof client.on !== "function") {
    logger.logger.warn("[Spotlight] Cannot connect to spotlight due to missing method on SDK client (`client.on`)");
    return;
  }
  client.on("beforeEnvelope", (envelope$1) => {
    if (failedRequests > 3) {
      logger.logger.warn("[Spotlight] Disabled Sentry -> Spotlight integration due to too many failed requests");
      return;
    }
    const serializedEnvelope = envelope.serializeEnvelope(envelope$1);
    const request = getNativeHttpRequest();
    const req = request(
      {
        method: "POST",
        path: spotlightUrl.pathname,
        hostname: spotlightUrl.hostname,
        port: spotlightUrl.port,
        headers: {
          "Content-Type": "application/x-sentry-envelope"
        }
      },
      (res) => {
        res.on("data", () => {
        });
        res.on("end", () => {
        });
        res.setEncoding("utf8");
      }
    );
    req.on("error", () => {
      failedRequests++;
      logger.logger.warn("[Spotlight] Failed to send envelope to Spotlight Sidecar");
    });
    req.write(serializedEnvelope);
    req.end();
  });
}
function parseSidecarUrl(url$1) {
  try {
    return new url.URL(`${url$1}`);
  } catch (e) {
    logger.logger.warn(`[Spotlight] Invalid sidecar URL: ${url$1}`);
    return void 0;
  }
}
function getNativeHttpRequest() {
  const { request } = http__namespace;
  if (isWrapped(request)) {
    return request.__sentry_original__;
  }
  return request;
}
function isWrapped(impl) {
  return "__sentry_original__" in impl;
}
exports.getNativeHttpRequest = getNativeHttpRequest;
exports.spotlightIntegration = spotlightIntegration;
