// @slicemachine/plugin-kit/dist/fs/index.js
import { checkIsTypeScriptProject } from "./checkIsTypeScriptProject.js";
import { upsertGlobalTypeScriptTypes } from "./upsertGlobalTypeScriptTypes.js";
import { checkHasProjectFile } from "./checkHasProjectFile.js";
import { writeProjectFile } from "./writeProjectFile.js";
import { readProjectFile } from "./readProjectFile.js";
import { deleteProjectFile } from "./deleteProjectFile.js";
import { readProjectEnvironment } from "./readProjectEnvironment.js";
import { writeProjectEnvironment } from "./writeProjectEnvironment.js";
import { buildSliceDirectoryPath } from "./buildSliceDirectoryPath.js";
import { buildSliceFilePath } from "./buildSliceFilePath.js";
import { buildSliceLibraryDirectoryPath } from "./buildSliceLibraryDirectoryPath.js";
import { readSliceLibrary } from "./readSliceLibrary.js";
import { writeSliceFile } from "./writeSliceFile.js";
import { writeSliceModel } from "./writeSliceModel.js";
import { readSliceFile } from "./readSliceFile.js";
import { readSliceModel } from "./readSliceModel.js";
import { renameSlice } from "./renameSlice.js";
import { deleteSliceFile } from "./deleteSliceFile.js";
import { deleteSliceDirectory } from "./deleteSliceDirectory.js";
import { buildCustomTypeDirectoryPath } from "./buildCustomTypeDirectoryPath.js";
import { buildCustomTypeFilePath } from "./buildCustomTypeFilePath.js";
import { buildCustomTypeLibraryDirectoryPath } from "./buildCustomTypeLibraryDirectoryPath.js";
import { readCustomTypeLibrary } from "./readCustomTypeLibrary.js";
import { writeCustomTypeFile } from "./writeCustomTypeFile.js";
import { writeCustomTypeModel } from "./writeCustomTypeModel.js";
import { readCustomTypeFile } from "./readCustomTypeFile.js";
import { readCustomTypeModel } from "./readCustomTypeModel.js";
import { renameCustomType } from "./renameCustomType.js";
import { deleteCustomTypeFile } from "./deleteCustomTypeFile.js";
import { deleteCustomTypeDirectory } from "./deleteCustomTypeDirectory.js";
import { readSliceTemplateLibrary } from "./readSliceTemplateLibrary.js";
export {
  buildCustomTypeDirectoryPath,
  buildCustomTypeFilePath,
  buildCustomTypeLibraryDirectoryPath,
  buildSliceDirectoryPath,
  buildSliceFilePath,
  buildSliceLibraryDirectoryPath,
  checkHasProjectFile,
  checkIsTypeScriptProject,
  deleteCustomTypeDirectory,
  deleteCustomTypeFile,
  deleteProjectFile,
  deleteSliceDirectory,
  deleteSliceFile,
  readCustomTypeFile,
  readCustomTypeLibrary,
  readCustomTypeModel,
  readProjectEnvironment,
  readProjectFile,
  readSliceFile,
  readSliceLibrary,
  readSliceModel,
  readSliceTemplateLibrary,
  renameCustomType,
  renameSlice,
  upsertGlobalTypeScriptTypes,
  writeCustomTypeFile,
  writeCustomTypeModel,
  writeProjectEnvironment,
  writeProjectFile,
  writeSliceFile,
  writeSliceModel
};
