// @slicemachine/plugin-kit/dist/fs/readSliceLibrary.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const path = require("node:path");
const checkPathExists = require("./lib/checkPathExists.cjs");
const isSharedSliceModel = require("./lib/isSharedSliceModel.cjs");
const readJSONFile = require("./lib/readJSONFile.cjs");
const fsLimit = require("./lib/fsLimit.cjs");
const constants = require("./constants.cjs");
const buildSliceLibraryDirectoryPath = require("./buildSliceLibraryDirectoryPath.cjs");
function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const path__namespace = /* @__PURE__ */ _interopNamespaceDefault(path);
const readSliceLibrary = async (args) => {
  const libraryDir = buildSliceLibraryDirectoryPath.buildSliceLibraryDirectoryPath({
    libraryID: args.libraryID,
    absolute: true,
    helpers: args.helpers
  });
  if (!await checkPathExists.checkPathExists(libraryDir)) {
    return {
      id: args.libraryID,
      sliceIDs: []
    };
  }
  const childDirs = await fsLimit.readdir(libraryDir, { withFileTypes: true });
  const unreadableModelPaths = [];
  const sliceIDs = [];
  await Promise.all(childDirs.map(async (childDir) => {
    if (childDir.isDirectory()) {
      const childDirContents = await fsLimit.readdir(path__namespace.join(libraryDir, childDir.name), {
        withFileTypes: true
      });
      const isSliceDir = childDirContents.some((entry) => {
        return entry.isFile() && entry.name === constants.SHARED_SLICE_MODEL_FILENAME;
      });
      if (isSliceDir) {
        const modelPath = path__namespace.join(libraryDir, childDir.name, constants.SHARED_SLICE_MODEL_FILENAME);
        try {
          const modelContents = await readJSONFile.readJSONFile(modelPath);
          if (isSharedSliceModel.isSharedSliceModel(modelContents)) {
            sliceIDs.push(modelContents.id);
          }
        } catch {
          unreadableModelPaths.push(modelPath);
        }
      }
    }
  }));
  if (unreadableModelPaths.length > 0) {
    const formattedPaths = unreadableModelPaths.join(", ");
    throw new Error(`The following Slice models could not be read: [${formattedPaths}]`);
  }
  return {
    id: args.libraryID,
    sliceIDs: sliceIDs.sort()
  };
};
exports.readSliceLibrary = readSliceLibrary;
