// @slicemachine/plugin-kit/dist/fs/upsertGlobalTypeScriptTypes.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const prismicTsCodegen = require("prismic-ts-codegen");
const writeProjectFile = require("./writeProjectFile.cjs");
const constants = require("./constants.cjs");
const upsertGlobalTypeScriptTypes = async (args) => {
  const project = await args.helpers.getProject();
  const [customTypeModelDescriptors, sharedSliceModelDescriptors] = await Promise.all([
    args.actions.readAllCustomTypeModels(),
    args.actions.readAllSliceModels()
  ]);
  const customTypeModels = customTypeModelDescriptors.map((customTypeModelDescriptors2) => {
    return customTypeModelDescriptors2.model;
  });
  const sharedSliceModels = sharedSliceModelDescriptors.map((sharedSliceModelDescriptors2) => {
    return sharedSliceModelDescriptors2.model;
  });
  let contents = prismicTsCodegen.generateTypes({
    customTypeModels,
    sharedSliceModels,
    clientIntegration: {
      includeCreateClientInterface: true,
      includeContentNamespace: true
    },
    typesProvider: await prismicTsCodegen.detectTypesProvider({ cwd: project.root })
  });
  contents = `${constants.NON_EDITABLE_FILE_BANNER}

${contents}`;
  return await writeProjectFile.writeProjectFile({
    ...args,
    filename: args.filename || constants.GLOBAL_TYPESCRIPT_TYPES_FILENAME,
    contents
  });
};
exports.upsertGlobalTypeScriptTypes = upsertGlobalTypeScriptTypes;
