// @slicemachine/manager/dist/lib/installDependencies.js
import { parseNi } from "@antfu/ni";
import { execaCommand } from './../_node_modules/execa/index.js';
const EXTRA_INSTALL_FLAGS = {
  npm: ["--color=always", "--loglevel=info"],
  pnpm: [],
  yarn: [],
  "yarn@berry": [],
  "pnpm@6": [],
  bun: []
  // TODO: Add the correct options.
};
const installDependencies = async (args) => {
  const commandArgs = Object.entries(args.dependencies).map(([pkg, range]) => `${pkg}@${range}`);
  if (commandArgs.length && args.dev) {
    commandArgs.unshift("-D");
  }
  commandArgs.push(...EXTRA_INSTALL_FLAGS[args.packageManager]);
  const command = await parseNi(args.packageManager, commandArgs);
  if (!command) {
    throw new Error("Failed to begin dependency installation (could not parse command)", {
      cause: {
        packageManager: args.packageManager,
        dependencies: args.dependencies
      }
    });
  }
  const execaProcess = execaCommand(command, {
    encoding: "utf-8",
    ...args.execa
  });
  return {
    execaProcess
  };
};
export {
  installDependencies
};
