// spdx-expression-parse/scan.js
import require$$1 from "../spdx-license-ids/index.json.js";
import require$$1$1 from "../spdx-license-ids/deprecated.json.js";
import require$$2 from "../spdx-exceptions/index.json.js";
var licenses = [].concat(require$$1).concat(require$$1$1);
var exceptions = require$$2;
var scan = function(source) {
  var index = 0;
  function hasMore() {
    return index < source.length;
  }
  function read(value) {
    if (value instanceof RegExp) {
      var chars = source.slice(index);
      var match = chars.match(value);
      if (match) {
        index += match[0].length;
        return match[0];
      }
    } else {
      if (source.indexOf(value, index) === index) {
        index += value.length;
        return value;
      }
    }
  }
  function skipWhitespace() {
    read(/[ ]*/);
  }
  function operator() {
    var string;
    var possibilities = ["WITH", "AND", "OR", "(", ")", ":", "+"];
    for (var i = 0; i < possibilities.length; i++) {
      string = read(possibilities[i]);
      if (string) {
        break;
      }
    }
    if (string === "+" && index > 1 && source[index - 2] === " ") {
      throw new Error("Space before `+`");
    }
    return string && {
      type: "OPERATOR",
      string
    };
  }
  function idstring() {
    return read(/[A-Za-z0-9-.]+/);
  }
  function expectIdstring() {
    var string = idstring();
    if (!string) {
      throw new Error("Expected idstring at offset " + index);
    }
    return string;
  }
  function documentRef() {
    if (read("DocumentRef-")) {
      var string = expectIdstring();
      return { type: "DOCUMENTREF", string };
    }
  }
  function licenseRef() {
    if (read("LicenseRef-")) {
      var string = expectIdstring();
      return { type: "LICENSEREF", string };
    }
  }
  function identifier() {
    var begin = index;
    var string = idstring();
    if (licenses.indexOf(string) !== -1) {
      return {
        type: "LICENSE",
        string
      };
    } else if (exceptions.indexOf(string) !== -1) {
      return {
        type: "EXCEPTION",
        string
      };
    }
    index = begin;
  }
  function parseToken() {
    return operator() || documentRef() || licenseRef() || identifier();
  }
  var tokens = [];
  while (hasMore()) {
    skipWhitespace();
    if (!hasMore()) {
      break;
    }
    var token = parseToken();
    if (!token) {
      throw new Error("Unexpected `" + source[index] + "` at offset " + index);
    }
    tokens.push(token);
  }
  return tokens;
};
export {
  scan as s
};
