// @prismicio/simulator/dist/index.js
import { SimulatorClient, simulatorClientDefaultOptions } from "./SimulatorClient.js";
import { SimulatorAPI, simulatorAPIDefaultOptions } from "./SimulatorAPI.js";
import { ChannelNotSetError, ConnectionTimeoutError, NotReadyError, PortNotSetError, RequestTimeoutError, ResponseError, TooManyConcurrentRequestsError } from "./channel/errors.js";
import { isErrorResponseMessage, isRequestMessage, isResponseMessage, isSuccessResponseMessage, validateMessage } from "./channel/messages.js";
import "./channel/types.js";
import { APIRequestType, ClientRequestType } from "./types.js";
export {
  APIRequestType,
  ChannelNotSetError,
  ClientRequestType,
  ConnectionTimeoutError,
  NotReadyError,
  PortNotSetError,
  RequestTimeoutError,
  ResponseError,
  SimulatorAPI,
  SimulatorClient,
  TooManyConcurrentRequestsError,
  isErrorResponseMessage,
  isRequestMessage,
  isResponseMessage,
  isSuccessResponseMessage,
  simulatorAPIDefaultOptions,
  simulatorClientDefaultOptions,
  validateMessage
};
