// @prismicio/simulator/dist/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const SimulatorClient = require("./SimulatorClient.cjs");
const SimulatorAPI = require("./SimulatorAPI.cjs");
const errors = require("./channel/errors.cjs");
const messages = require("./channel/messages.cjs");
require("./channel/types.cjs");
const types = require("./types.cjs");
exports.SimulatorClient = SimulatorClient.SimulatorClient;
exports.simulatorClientDefaultOptions = SimulatorClient.simulatorClientDefaultOptions;
exports.SimulatorAPI = SimulatorAPI.SimulatorAPI;
exports.simulatorAPIDefaultOptions = SimulatorAPI.simulatorAPIDefaultOptions;
exports.ChannelNotSetError = errors.ChannelNotSetError;
exports.ConnectionTimeoutError = errors.ConnectionTimeoutError;
exports.NotReadyError = errors.NotReadyError;
exports.PortNotSetError = errors.PortNotSetError;
exports.RequestTimeoutError = errors.RequestTimeoutError;
exports.ResponseError = errors.ResponseError;
exports.TooManyConcurrentRequestsError = errors.TooManyConcurrentRequestsError;
exports.isErrorResponseMessage = messages.isErrorResponseMessage;
exports.isRequestMessage = messages.isRequestMessage;
exports.isResponseMessage = messages.isResponseMessage;
exports.isSuccessResponseMessage = messages.isSuccessResponseMessage;
exports.validateMessage = messages.validateMessage;
Object.defineProperty(exports, "APIRequestType", {
  enumerable: true,
  get: () => types.APIRequestType
});
Object.defineProperty(exports, "ClientRequestType", {
  enumerable: true,
  get: () => types.ClientRequestType
});
