// @prismicio/simulator/dist/kit.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const SimulatorManager = require("./kit/SimulatorManager.cjs");
const State = require("./kit/State.cjs");
const getDefault = require("./kit/getDefault.cjs");
const domHelpers = require("./kit/domHelpers.cjs");
const eventHandlers = require("./kit/eventHandlers.cjs");
const messages = require("./kit/messages.cjs");
const types = require("./kit/types.cjs");
exports.SimulatorManager = SimulatorManager.SimulatorManager;
exports.State = State.State;
exports.getDefaultMessage = getDefault.getDefaultMessage;
exports.getDefaultProps = getDefault.getDefaultProps;
exports.getDefaultSlices = getDefault.getDefaultSlices;
exports.getActiveSliceDOM = domHelpers.getActiveSliceDOM;
exports.getSliceZoneDOM = domHelpers.getSliceZoneDOM;
exports.simulatorClass = domHelpers.simulatorClass;
exports.simulatorRootClass = domHelpers.simulatorRootClass;
exports.disableEventHandler = eventHandlers.disableEventHandler;
exports.onClickHandler = eventHandlers.onClickHandler;
exports.sliceSimulatorAccessedDirectly = messages.sliceSimulatorAccessedDirectly;
Object.defineProperty(exports, "StateEventType", {
  enumerable: true,
  get: () => types.StateEventType
});
