// @prismicio/vue/dist/index.js
import { createPrismic } from "./createPrismic.js";
import { usePrismic } from "./usePrismic.js";
import { dangerouslyUseAllPrismicDocuments, useAllPrismicDocumentsByEveryTag, useAllPrismicDocumentsByIDs, useAllPrismicDocumentsBySomeTags, useAllPrismicDocumentsByTag, useAllPrismicDocumentsByType, useAllPrismicDocumentsByUIDs, useFirstPrismicDocument, usePrismicDocumentByID, usePrismicDocumentByUID, usePrismicDocuments, usePrismicDocumentsByEveryTag, usePrismicDocumentsByIDs, usePrismicDocumentsBySomeTags, usePrismicDocumentsByTag, usePrismicDocumentsByType, usePrismicDocumentsByUIDs, useSinglePrismicDocument } from "./composables.js";
import { prismicKey } from "./injectionSymbols.js";
import { PrismicImage, usePrismicImage } from "./components/PrismicImage.js";
import { PrismicLink, usePrismicLink } from "./components/PrismicLink.js";
import { PrismicText, usePrismicText } from "./components/PrismicText.js";
import { PrismicRichText, usePrismicRichText } from "./components/PrismicRichText.js";
import { PrismicEmbed } from "./components/PrismicEmbed.js";
import { SliceZone, TODOSliceComponent, defineSliceZoneComponents, getSliceComponentProps } from "./components/SliceZone.js";
export {
  PrismicEmbed,
  PrismicImage,
  PrismicLink,
  PrismicRichText,
  PrismicText,
  SliceZone,
  TODOSliceComponent,
  createPrismic,
  dangerouslyUseAllPrismicDocuments,
  defineSliceZoneComponents,
  getSliceComponentProps,
  prismicKey,
  useAllPrismicDocumentsByEveryTag,
  useAllPrismicDocumentsByIDs,
  useAllPrismicDocumentsBySomeTags,
  useAllPrismicDocumentsByTag,
  useAllPrismicDocumentsByType,
  useAllPrismicDocumentsByUIDs,
  useFirstPrismicDocument,
  usePrismic,
  usePrismicDocumentByID,
  usePrismicDocumentByUID,
  usePrismicDocuments,
  usePrismicDocumentsByEveryTag,
  usePrismicDocumentsByIDs,
  usePrismicDocumentsBySomeTags,
  usePrismicDocumentsByTag,
  usePrismicDocumentsByType,
  usePrismicDocumentsByUIDs,
  usePrismicImage,
  usePrismicLink,
  usePrismicRichText,
  usePrismicText,
  useSinglePrismicDocument
};
