// @prismicio/api-renderer/lib/models/customtypes/fields/slices/CompositeSliceDef.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompositeSliceDef = void 0;
const tslib_1 = require("tslib");
const customtypes_1 = require("@prismicio/types-internal/lib/customtypes");
const fp_ts_1 = require("fp-ts");
const function_1 = require("fp-ts/lib/function");
const t = tslib_1.__importStar(require("io-ts"));
const utils_1 = require("../../utils");
const GroupDef_1 = require("../GroupDef");
const NestableDef_1 = require("../nestable/NestableDef");
const sliceCodec = t.strict({
    name: t.string,
    primary: t.array(NestableDef_1.NestableDef),
    items: GroupDef_1.NestedGroupDef,
});
exports.CompositeSliceDef = new t.Type("CompositeSliceDef", (u) => (0, utils_1.isFieldDef)(u) && customtypes_1.CompositeSlice.is(u.def), (u) => {
    return (0, function_1.pipe)(sliceCodec.decode(u), fp_ts_1.either.map(({ name, primary, items }) => {
        var _a;
        return ({
            key: name,
            def: {
                type: "Slice",
                "non-repeat": primary.reduce((acc, { key, def }) => ({ ...acc, [key]: def }), {}),
                repeat: (_a = items.def.config) === null || _a === void 0 ? void 0 : _a.fields,
            },
        });
    }));
}, (s) => s);
