// @prismicio/api-renderer/lib/models/customtypes/fields/slices/SlicesDef.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlicesDef = void 0;
const tslib_1 = require("tslib");
const customtypes_1 = require("@prismicio/types-internal/lib/customtypes");
const fp_ts_1 = require("fp-ts");
const function_1 = require("fp-ts/lib/function");
const t = tslib_1.__importStar(require("io-ts"));
const utils_1 = require("../../utils");
const SliceDef_1 = require("./SliceDef");
const codec = t.strict({
    name: t.string,
    slices: t.array(SliceDef_1.SliceDef),
});
exports.SlicesDef = new t.Type("SlicesDef", (u) => (0, utils_1.isFieldDef)(u) && customtypes_1.StaticSlices.is(u.def), (u) => {
    return (0, function_1.pipe)(codec.decode(u), fp_ts_1.either.map(({ name, slices }) => {
        return {
            key: name,
            def: {
                type: "Slices",
                config: {
                    choices: slices.reduce((acc, { key, def }) => ({ ...acc, [key]: def }), {}),
                },
            },
        };
    }));
}, (s) => s);
