// @prismicio/api-renderer/src/widgets/nestable/Link/FileLinkRenderer.ts
import type { FileLinkContent } from "@prismicio/types-internal/lib/content"

import type { RenderContext } from "../../../models"
import type { AnyLinkDef, LinkRenderer, MediaLinkDef } from "../../../models"

const FileLinkRenderer: (ctx: RenderContext) => LinkRenderer<MediaLinkDef, FileLinkContent> = (ctx) => ({
	renderV1(link: FileLinkContent): unknown {
		return {
			type: "Link.file",
			value: {
				file: {
					name: link.name,
					kind: link.kind,
					url: ctx.urlRewriter.rewriteFileUrl(link.url),
					size: link.size,
				},
			},
		}
	},

	renderV2(_def: MediaLinkDef | AnyLinkDef, link: FileLinkContent): unknown {
		return {
			link_type: "Media",
			id: link.id,
			name: link.name,
			kind: link.kind,
			url: ctx.urlRewriter.rewriteFileUrl(link.url),
			size: link.size,
		}
	},

	renderMocks(_def: MediaLinkDef | AnyLinkDef, _link: FileLinkContent): unknown {
		return {
			link_type: "Media",
			name: "mock",
			kind: "image",
			url: "https://source.unsplash.com/random",
			size: "10kB",
		}
	},
})

export default FileLinkRenderer
