// @slicemachine/manager/dist/client.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const createSliceMachineManagerClient = require("./managers/createSliceMachineManagerClient.cjs");
const GIT_PROVIDER = require("./constants/GIT_PROVIDER.cjs");
const errors = require("./errors.cjs");
const DecodeError = require("./lib/DecodeError.cjs");
exports.createSliceMachineManagerClient = createSliceMachineManagerClient.createSliceMachineManagerClient;
exports.GIT_PROVIDER = GIT_PROVIDER.GIT_PROVIDER;
exports.InternalError = errors.InternalError;
exports.NotFoundError = errors.NotFoundError;
exports.SliceMachineError = errors.SliceMachineError;
exports.UnauthenticatedError = errors.UnauthenticatedError;
exports.UnauthorizedError = errors.UnauthorizedError;
exports.UnexpectedDataError = errors.UnexpectedDataError;
exports.isInternalError = errors.isInternalError;
exports.isInvalidActiveEnvironmentError = errors.isInvalidActiveEnvironmentError;
exports.isNotFoundError = errors.isNotFoundError;
exports.isPluginError = errors.isPluginError;
exports.isSliceMachineError = errors.isSliceMachineError;
exports.isUnauthenticatedError = errors.isUnauthenticatedError;
exports.isUnauthorizedError = errors.isUnauthorizedError;
exports.isUnexpectedDataError = errors.isUnexpectedDataError;
exports.isUnsupportedError = errors.isUnsupportedError;
exports.DecodeError = DecodeError.DecodeError;
