// @slicemachine/manager/dist/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const createSliceMachineManager = require("./managers/createSliceMachineManager.cjs");
const createSliceMachineManagerMiddleware = require("./managers/createSliceMachineManagerMiddleware.cjs");
const createPrismicAuthManager = require("./auth/createPrismicAuthManager.cjs");
const createPrismicAuthManagerMiddleware = require("./auth/createPrismicAuthManagerMiddleware.cjs");
const errors = require("./errors.cjs");
const getEnvironmentInfo = require("./getEnvironmentInfo.cjs");
const DecodeError = require("./lib/DecodeError.cjs");
const REPOSITORY_NAME_VALIDATION = require("./constants/REPOSITORY_NAME_VALIDATION.cjs");
exports.createSliceMachineManager = createSliceMachineManager.createSliceMachineManager;
exports.createSliceMachineManagerMiddleware = createSliceMachineManagerMiddleware.createSliceMachineManagerMiddleware;
exports.createPrismicAuthManager = createPrismicAuthManager.createPrismicAuthManager;
exports.createPrismicAuthManagerMiddleware = createPrismicAuthManagerMiddleware.createPrismicAuthManagerMiddleware;
exports.InternalError = errors.InternalError;
exports.InvalidActiveEnvironmentError = errors.InvalidActiveEnvironmentError;
exports.NotFoundError = errors.NotFoundError;
exports.PluginError = errors.PluginError;
exports.PluginHookResultError = errors.PluginHookResultError;
exports.SliceMachineError = errors.SliceMachineError;
exports.UnauthenticatedError = errors.UnauthenticatedError;
exports.UnauthorizedError = errors.UnauthorizedError;
exports.UnexpectedDataError = errors.UnexpectedDataError;
exports.UnsupportedError = errors.UnsupportedError;
exports.getEnvironmentInfo = getEnvironmentInfo.getEnvironmentInfo;
exports.DecodeError = DecodeError.DecodeError;
exports.REPOSITORY_NAME_VALIDATION = REPOSITORY_NAME_VALIDATION.REPOSITORY_NAME_VALIDATION;
