// @slicemachine/manager/dist/index.js
import { createSliceMachineManager } from "./managers/createSliceMachineManager.js";
import { createSliceMachineManagerMiddleware } from "./managers/createSliceMachineManagerMiddleware.js";
import { createPrismicAuthManager } from "./auth/createPrismicAuthManager.js";
import { createPrismicAuthManagerMiddleware } from "./auth/createPrismicAuthManagerMiddleware.js";
import { InternalError, InvalidActiveEnvironmentError, NotFoundError, PluginError, PluginHookResultError, SliceMachineError, UnauthenticatedError, UnauthorizedError, UnexpectedDataError, UnsupportedError } from "./errors.js";
import { getEnvironmentInfo } from "./getEnvironmentInfo.js";
import { DecodeError } from "./lib/DecodeError.js";
import { REPOSITORY_NAME_VALIDATION } from "./constants/REPOSITORY_NAME_VALIDATION.js";
export {
  DecodeError,
  InternalError,
  InvalidActiveEnvironmentError,
  NotFoundError,
  PluginError,
  PluginHookResultError,
  REPOSITORY_NAME_VALIDATION,
  SliceMachineError,
  UnauthenticatedError,
  UnauthorizedError,
  UnexpectedDataError,
  UnsupportedError,
  createPrismicAuthManager,
  createPrismicAuthManagerMiddleware,
  createSliceMachineManager,
  createSliceMachineManagerMiddleware,
  getEnvironmentInfo
};
