// @amplitude/experiment-node-server/dist/src/index.js
import { __exports as src } from "../../../../../_virtual/index.js";
import "./cookie.js";
import "./remote/client.js";
import "./types/config.js";
import "./factory.js";
import "./local/client.js";
import "./local/fetcher.js";
import "./local/poller.js";
import "./local/cache.js";
import { __exports as cookie } from "../../../../../_virtual/cookie.js";
import { __exports as client } from "../../../../../_virtual/client.js";
import { __exports as config } from "../../../../../_virtual/config.js";
import { __exports as factory } from "../../../../../_virtual/factory.js";
import { __exports as client$1 } from "../../../../../_virtual/client2.js";
import { __exports as fetcher } from "../../../../../_virtual/fetcher.js";
import { __exports as poller } from "../../../../../_virtual/poller.js";
import { __exports as cache } from "../../../../../_virtual/cache.js";
(function(exports) {
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.InMemoryFlagConfigCache = exports.FlagConfigPoller = exports.FlagConfigFetcher = exports.LocalEvaluationClient = exports.Experiment = exports.RemoteEvaluationDefaults = exports.Defaults = exports.RemoteEvaluationClient = exports.ExperimentClient = exports.AmplitudeCookie = void 0;
  var cookie_1 = cookie;
  Object.defineProperty(exports, "AmplitudeCookie", { enumerable: true, get: function() {
    return cookie_1.AmplitudeCookie;
  } });
  var client_1 = client;
  Object.defineProperty(exports, "ExperimentClient", { enumerable: true, get: function() {
    return client_1.ExperimentClient;
  } });
  Object.defineProperty(exports, "RemoteEvaluationClient", { enumerable: true, get: function() {
    return client_1.RemoteEvaluationClient;
  } });
  var config_1 = config;
  Object.defineProperty(exports, "Defaults", { enumerable: true, get: function() {
    return config_1.Defaults;
  } });
  Object.defineProperty(exports, "RemoteEvaluationDefaults", { enumerable: true, get: function() {
    return config_1.RemoteEvaluationDefaults;
  } });
  var factory_1 = factory;
  Object.defineProperty(exports, "Experiment", { enumerable: true, get: function() {
    return factory_1.Experiment;
  } });
  var client_2 = client$1;
  Object.defineProperty(exports, "LocalEvaluationClient", { enumerable: true, get: function() {
    return client_2.LocalEvaluationClient;
  } });
  var fetcher_1 = fetcher;
  Object.defineProperty(exports, "FlagConfigFetcher", { enumerable: true, get: function() {
    return fetcher_1.FlagConfigFetcher;
  } });
  var poller_1 = poller;
  Object.defineProperty(exports, "FlagConfigPoller", { enumerable: true, get: function() {
    return poller_1.FlagConfigPoller;
  } });
  var cache_1 = cache;
  Object.defineProperty(exports, "InMemoryFlagConfigCache", { enumerable: true, get: function() {
    return cache_1.InMemoryFlagConfigCache;
  } });
})(src);
