// @amplitude/analytics-types/lib/esm/event.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
exports.IdentifyOperation = void 0;
(function(IdentifyOperation) {
  IdentifyOperation["SET"] = "$set";
  IdentifyOperation["SET_ONCE"] = "$setOnce";
  IdentifyOperation["ADD"] = "$add";
  IdentifyOperation["APPEND"] = "$append";
  IdentifyOperation["PREPEND"] = "$prepend";
  IdentifyOperation["REMOVE"] = "$remove";
  IdentifyOperation["PREINSERT"] = "$preInsert";
  IdentifyOperation["POSTINSERT"] = "$postInsert";
  IdentifyOperation["UNSET"] = "$unset";
  IdentifyOperation["CLEAR_ALL"] = "$clearAll";
})(exports.IdentifyOperation || (exports.IdentifyOperation = {}));
exports.RevenueProperty = void 0;
(function(RevenueProperty) {
  RevenueProperty["REVENUE_PRODUCT_ID"] = "$productId";
  RevenueProperty["REVENUE_QUANTITY"] = "$quantity";
  RevenueProperty["REVENUE_PRICE"] = "$price";
  RevenueProperty["REVENUE_TYPE"] = "$revenueType";
  RevenueProperty["REVENUE"] = "$revenue";
})(exports.RevenueProperty || (exports.RevenueProperty = {}));
exports.SpecialEventType = void 0;
(function(SpecialEventType) {
  SpecialEventType["IDENTIFY"] = "$identify";
  SpecialEventType["GROUP_IDENTIFY"] = "$groupidentify";
  SpecialEventType["REVENUE"] = "revenue_amount";
})(exports.SpecialEventType || (exports.SpecialEventType = {}));
