// @amplitude/analytics-types/lib/esm/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const event = require("./event.cjs");
const logger = require("./logger.cjs");
const plugin = require("./plugin.cjs");
const serverZone = require("./server-zone.cjs");
const status = require("./status.cjs");
const transport = require("./transport.cjs");
Object.defineProperty(exports, "IdentifyOperation", {
  enumerable: true,
  get: () => event.IdentifyOperation
});
Object.defineProperty(exports, "RevenueProperty", {
  enumerable: true,
  get: () => event.RevenueProperty
});
Object.defineProperty(exports, "SpecialEventType", {
  enumerable: true,
  get: () => event.SpecialEventType
});
Object.defineProperty(exports, "LogLevel", {
  enumerable: true,
  get: () => logger.LogLevel
});
Object.defineProperty(exports, "PluginType", {
  enumerable: true,
  get: () => plugin.PluginType
});
Object.defineProperty(exports, "ServerZone", {
  enumerable: true,
  get: () => serverZone.ServerZone
});
Object.defineProperty(exports, "Status", {
  enumerable: true,
  get: () => status.Status
});
Object.defineProperty(exports, "TransportType", {
  enumerable: true,
  get: () => transport.TransportType
});
