// @prismicio/mocks/lib/generators/widgets/nestable/Color.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColorMock = void 0;
const content_1 = require("@prismicio/types-internal/lib/content");
const content_2 = require("@prismicio/types-internal/lib/content");
const diff_1 = require("@prismicio/types-internal/lib/customtypes/diff");
function random() {
    const hexLetters = "0123456789abcdef";
    let color = "#";
    for (let i = 0; i < 6; i++) {
        color += hexLetters[Math.floor(Math.random() * 16)];
    }
    return color;
}
exports.ColorMock = {
    generate(_def, config) {
        const value = (config === null || config === void 0 ? void 0 : config.value) || random();
        return {
            __TYPE__: content_2.FieldContentType,
            value,
            type: "Color",
        };
    },
    applyPatch(data) {
        var _a;
        if (data.diff.op === diff_1.DiffOperation.Removed)
            return { result: undefined };
        if (data.diff.value.type === "Color") {
            const patched = this.patch(data.diff, (0, content_1.isColorContent)(data.content) ? data.content : undefined, ((_a = data.config) === null || _a === void 0 ? void 0 : _a.type) === "Color" ? data.config : undefined);
            return { result: patched };
        }
        return;
    },
    patch(diff, _content, config) {
        switch (diff.op) {
            case diff_1.DiffOperation.Removed:
                return;
            case diff_1.DiffOperation.Updated:
            case diff_1.DiffOperation.Added:
                return this.generate(diff.value, config);
        }
    },
};
