// @prismicio/mocks/lib/generators/widgets/nestable/Date.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DateMock = void 0;
const content_1 = require("@prismicio/types-internal/lib/content");
const diff_1 = require("@prismicio/types-internal/lib/customtypes/diff");
function formatDate(date) {
    const formatted = date.toISOString().split("T")[0];
    if (!formatted)
        throw new Error("Something happened during Date generation.");
    return formatted;
}
function random(start = new Date(2012, 0, 1), end = new Date()) {
    const genDate = new Date(start.getTime() + Math.random() * (end.getTime() - start.getTime()));
    return formatDate(genDate);
}
exports.DateMock = {
    generate(_def, config) {
        const value = (() => {
            if (config === null || config === void 0 ? void 0 : config.value)
                return formatDate(config.value);
            if ((config === null || config === void 0 ? void 0 : config.start) && (config === null || config === void 0 ? void 0 : config.end))
                return random(config.start, config.end);
            return random(new Date(2012, 0, 1), new Date());
        })();
        return {
            __TYPE__: content_1.FieldContentType,
            type: "Date",
            value,
        };
    },
    applyPatch(data) {
        var _a;
        if (data.diff.op === diff_1.DiffOperation.Removed)
            return { result: undefined };
        if (data.diff.value.type === "Date") {
            const patched = this.patch(data.diff, (0, content_1.isDateContent)(data.content) ? data.content : undefined, ((_a = data.config) === null || _a === void 0 ? void 0 : _a.type) === "Date" ? data.config : undefined);
            return { result: patched };
        }
        return;
    },
    patch(diff, _content, config) {
        switch (diff.op) {
            case diff_1.DiffOperation.Removed:
                return;
            case diff_1.DiffOperation.Updated:
            case diff_1.DiffOperation.Added:
                return this.generate(diff.value, config);
        }
    },
};
