// @prismicio/vue/dist/components/PrismicEmbed.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const vue = require("vue");
const simplyResolveComponent = require("../lib/simplyResolveComponent.cjs");
const defaultWrapper = "div";
const PrismicEmbedImpl = /* @__PURE__ */ vue.defineComponent({
  name: "PrismicEmbed",
  props: {
    field: {
      type: Object,
      required: true
    },
    wrapper: {
      type: [String, Object, Function],
      default: void 0,
      required: false
    }
  },
  setup(props) {
    if (!props.field) {
      return () => null;
    }
    return () => {
      return vue.h(simplyResolveComponent.simplyResolveComponent(props.wrapper || defaultWrapper), {
        "data-oembed": props.field.embed_url,
        "data-oembed-type": props.field.type,
        "data-oembed-provider": props.field.provider_name,
        innerHTML: props.field.html || null
      });
    };
  }
});
const PrismicEmbed = PrismicEmbedImpl;
exports.PrismicEmbed = PrismicEmbed;
exports.PrismicEmbedImpl = PrismicEmbedImpl;
