// @prismicio/vue/dist/components/PrismicText.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const client = require("@prismicio/client");
const vue = require("vue");
const simplyResolveComponent = require("../lib/simplyResolveComponent.cjs");
const defaultWrapper = "div";
const usePrismicText = (props) => {
  const text = vue.computed(() => {
    const field = vue.unref(props.field);
    if (!client.isFilled.richText(field)) {
      return vue.unref(props.fallback) ?? "";
    }
    return client.asText(vue.unref(field), vue.unref(props.separator));
  });
  return {
    text
  };
};
const PrismicTextImpl = /* @__PURE__ */ vue.defineComponent({
  name: "PrismicText",
  props: {
    field: {
      type: Array,
      default: void 0,
      required: false
    },
    separator: {
      type: String,
      default: void 0,
      required: false
    },
    wrapper: {
      type: [String, Object, Function],
      default: void 0,
      required: false
    },
    fallback: {
      type: String,
      default: void 0,
      required: false
    }
  },
  setup(props) {
    const { text } = usePrismicText(props);
    return () => {
      const parent = simplyResolveComponent.simplyResolveComponent(props.wrapper || defaultWrapper);
      return vue.h(parent, null, {
        default: () => text.value
      });
    };
  }
});
const PrismicText = PrismicTextImpl;
exports.PrismicText = PrismicText;
exports.PrismicTextImpl = PrismicTextImpl;
exports.usePrismicText = usePrismicText;
