// @amplitude/analytics-node/lib/esm/plugins/context.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const tslib_es6 = require("../../../../../tslib/tslib.es6.cjs");
const version = require("../version.cjs");
const plugin = require("../../../../analytics-types/lib/esm/plugin.cjs");
const uuid = require("../../../../analytics-core/lib/esm/utils/uuid.cjs");
var Context = (
  /** @class */
  function() {
    function Context2() {
      this.name = "context";
      this.type = plugin.PluginType.BEFORE;
      this.eventId = 0;
      this.library = "amplitude-node-ts/".concat(version.VERSION);
    }
    Context2.prototype.setup = function(config) {
      this.config = config;
      return Promise.resolve(void 0);
    };
    Context2.prototype.execute = function(context) {
      var _this = this;
      return new Promise(function(resolve) {
        var time = (/* @__PURE__ */ new Date()).getTime();
        var contextEvent = tslib_es6.__assign(tslib_es6.__assign(tslib_es6.__assign({ time, insert_id: uuid.UUID(), plan: _this.config.plan }, _this.config.ingestionMetadata && {
          ingestion_metadata: {
            source_name: _this.config.ingestionMetadata.sourceName,
            source_version: _this.config.ingestionMetadata.sourceVersion
          }
        }), context), { event_id: _this.eventId++, library: _this.library });
        return resolve(contextEvent);
      });
    };
    return Context2;
  }()
);
exports.Context = Context;
