// @prismicio/types-internal/lib/content/fields/nestable/IntegrationFieldContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntegrationFieldContent = exports.IntegrationFieldLegacy = exports.isIntegrationFieldContent = exports.IntegrationFieldContentType = void 0;
const tslib_1 = require("tslib");
const fp_ts_1 = require("fp-ts");
const function_1 = require("fp-ts/lib/function");
const t = (0, tslib_1.__importStar)(require("io-ts"));
const utils_1 = require("../../utils");
exports.IntegrationFieldContentType = "IntegrationFieldsContent";
const isIntegrationFieldContent = (u) => (0, utils_1.hasContentType)(u) && u.__TYPE__ === exports.IntegrationFieldContentType;
exports.isIntegrationFieldContent = isIntegrationFieldContent;
const IntegrationFieldLegacy = (ctx) => new t.Type("IntegrationFieldLegacy", exports.isIntegrationFieldContent, (u) => {
    return (0, function_1.pipe)(t.string.decode(u), fp_ts_1.either.map((i) => exports.IntegrationFieldContent.encode({
        value: i,
        __TYPE__: exports.IntegrationFieldContentType,
    })));
}, (i) => {
    return {
        content: i.value,
        types: { [ctx.keyOfType]: "IntegrationFields" },
    };
});
exports.IntegrationFieldLegacy = IntegrationFieldLegacy;
exports.IntegrationFieldContent = t.strict({
    __TYPE__: t.literal(exports.IntegrationFieldContentType),
    value: t.string,
});
